smolyay/surrogate.py

```diff
@@ -0,0 +1,279 @@
+        self._make_basis_matrix()
+
+        # transform grid points
+        transformed_grids = numpy.zeros((
+            len(self._grids.points), self._dimension))
+        for dimension_ in range(self._dimension):
+            transformed_grids[:, dimension_] = (
+                numpy.polynomial.polyutils.mapdomain(
+                    numpy.array(self._grids.points)[:, dimension_],
+                    (-1, 1), self.domain[dimension_]))
+
+        # evaluate real function at grid points
+        real_function_at_grids = numpy.zeros((len(self._grids.points)))
+        for point_num in range(len(self._grids.points)):
+            real_function_at_grids[point_num] = (
+                self.real_function(*transformed_grids[point_num]))
+
+        # fit
+        self._coefficients = solve_linear(
+            self._basis_matrix, real_function_at_grids,
+            self.linear_solver)
+
+        return self._coefficients.tolist()
+
+    def train_from_data(self, real_function_at_grids, linear_solver='lu'):
+        """Fit surrogate's components (basis functions) to data.
+
+        Parameters
+        ----------
+        real_function_at_grids: list
+            Real function at grid points.
+        linear_slover: string
+            Method of solving linear equations:
+                'lu': lower upper decomposition
+                'inv': solve for coefficients through x = A^{-1}B.
+                'lstsq': least-square solution.
+
+        Returns
+        -------
+        list
+            Coefficients of the surrogate.
+
+        Raises
+        ------
+        IndexError
+            Data should be of length of the grid points.
+        """
+        if len(real_function_at_grids) != len(self._grids.points):
+            raise IndexError("Data must be in length of the grid points.")
+        self.real_function_at_grids = real_function_at_grids
+        self.linear_solver = linear_solver
+        self._make_basis_matrix()
+        self._coefficients = solve_linear(self._basis_matrix,
+                                          self.real_function_at_grids,
+                                          self.linear_solver)
+
+        return self._coefficients.tolist()
+
+    def __call__(self, x):
+        """Evaluate surrogate at a given input.
+
+        Parameter
+        ---------
+        x: list (for dimension > 1), or float (for dimension = 1)
+            Input.
+
+        Returns
+        -------
+        float
+            Surrogate output at x.
+
+        Raises
+        ------
+        ValueError
+            For surrogate to be evaluated, function needs to be trained.
+
+        IndexError
+            Input of the surrogate must be of length of the
+            real function's dimensionality.
+        """
+        if self._coefficients is None:
+            raise ValueError('Function needs training!')
+        if isinstance(x, (int, float)):
+            x = [x]
+        if len(x) != self._dimension:
```

Comment:
If you cast to ndarray, this would be a stricter test:
```suggestion
        if x.shape != (self._dimension,):
```